src/main/java/htsjdk/variant/variantcontext/VariantContext.java

```diff
@@ -1638,106 +1651,221 @@ public boolean isFullyDecoded() {
+            return this;
+        } else {
+            // This will ensure that the VariantContext is decoded in a percent-encoding aware way
+            final VariantContext newVC = this.fullyDecode(header, true);
+            newVC.version = headerVersion;
+
+            final Object oldGP = newVC.getAttribute(VCFConstants.GENOTYPE_POSTERIORS_KEY);
+            // We need to special-case GP because there is a discrepancy in the scale used to record
+            // its values between pre-4.3 and 4.3+ VCF. Pre-4.3 GP is phred scale encoded while
+            // 4.3+ GP is a linear probability, bringing it in line with other standard keys that
+            // use the P suffix (c.f. VCF 4.3 spec section 7.2).
+            if (headerVersion.isAtLeastAsRecentAs(VCFHeaderVersion.VCF4_3) && oldGP != null) {
+                if (oldGP instanceof List
+                    && (!((List<?>) oldGP).isEmpty() && ((List<?>) oldGP).get(0) instanceof Double)
+                ) {
+                    newVC.commonInfo.putAttribute(
+                        VCFConstants.GENOTYPE_POSTERIORS_KEY,
+                        makeGPValueLinear((List<Double>) oldGP),
+                        true
+                    );
                 }
             }
-        } catch (NumberFormatException e) {
-            throw new TribbleException("Could not decode field " + field + " with value " + string + " of declared type " + format.getType());
+            return newVC;
         }
     }
 
-    private final void fullyDecodeGenotypes(final VariantContextBuilder builder, final VCFHeader header) {
-        final GenotypesContext gc = new GenotypesContext();
-        for ( final Genotype g : getGenotypes() ) {
-            gc.add(fullyDecodeGenotypes(g, header));
+    private static List<Double> makeGPValueLinear(final List<Double> gpValues) {
+        // Some tools in the wild apparently already use linear scaled GP, so we have to
+        // be careful about converting inputs. We check whether GP values are already linear
+        // scaled by seeing if the values' sum is approximately equal to 1, like we
+        // would expect if the values were linear scale probabilities.
+        // c.f. https://sourceforge.net/p/vcftools/mailman/vcftools-spec/thread/CEBCD558.FA29%25browning%40u.washington.edu/
+        double sum = 0;
+        for (final double d : gpValues) {
+            sum += d;
         }
-        builder.genotypesNoValidation(gc);
-    }
 
-    private final Genotype fullyDecodeGenotypes(final Genotype g, final VCFHeader header) {
-        final Map<String, Object> map = fullyDecodeAttributes(g.getExtendedAttributes(), header, true);
-        return new GenotypeBuilder(g).attributes(map).make();
+        final boolean wasLinearScale = GeneralUtils.compareDoubles(sum, 1, VCFConstants.VCF_ENCODING_EPSILON) == 0;
+        if (!wasLinearScale) {
+            gpValues.replaceAll(GP -> QualityUtil.getErrorProbabilityFromPhredScore((int) Math.round(GP)));
+        }
```

Comment:
This should log a warning if it finds values that don't appear to be version-conforming (i.e, pre v4.3 and already linear).